### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by AlexandruGH**

### CODE REVIEW

File: Ghiurutan V. Alexandru/Homework5/src/model/Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+public class Dictionary implements Observer, DictionaryProc {
+	private static Dictionary instance;
+	private Map<Word, String> dictionary;
+	private Set<String> otherWords;
+	private Iterator<Word> it;
+	private Iterator<Entry<Word, String>> iterator;
+
+	private Dictionary() {
+		dictionary = new HashMap<Word, String>();
+
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/model/Dictionary.java:25]: I added some data for testing purposes.

File: Ghiurutan V. Alexandru/Homework5/src/model/DictionaryProc.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package model;
+
+public interface DictionaryProc {
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/model/DictionaryProc.java:3]: I didn't add the pre,post ,invariant yet.

File: Ghiurutan V. Alexandru/Homework5/src/model/Word.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Observable;
+
+public class Word extends Observable {
+
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/model/Word.java:8]: Here I use the Observer and Composite design patterns.
Composite is used effectively when adding a new word to the dictionary and Observer is used especially at deletion.Both ensure consistency of the dictionary.

File: Ghiurutan V. Alexandru/Homework5/src/view/Gui.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+					"Enter the description of the word: " + auxWord + ", that you want in dictionary.", "Input",
+					JOptionPane.INFORMATION_MESSAGE)).toLowerCase();
+		}
+		return description;
+	}
+
+	private void addWords(Word newWord) {
+		String description = checkInputDescription(newWord.getWord());
+		dictionary.addWord(newWord, description);
+		description = description.replaceAll("[^A-Za-z-]", " ");
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/view/Gui.java:123]: I ensure here that I don't have words with special characters or too much free spaces.

File: Ghiurutan V. Alexandru/Homework5/src/view/Gui.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+	private String checkInputDescription(String auxWord) {
+		String description = String.valueOf(JOptionPane.showInputDialog(this,
+				"Enter the description of the word: " + auxWord + ", that you want in dictionary.", "Input",
+				JOptionPane.INFORMATION_MESSAGE)).toLowerCase();
+		while (description.equals("")) {
+			JOptionPane.showMessageDialog(this, "You didn't enter a valid word.Please enter again more carefully.",
+					"ERROR", JOptionPane.ERROR_MESSAGE);
+			description = String.valueOf(JOptionPane.showInputDialog(this,
+					"Enter the description of the word: " + auxWord + ", that you want in dictionary.", "Input",
+					JOptionPane.INFORMATION_MESSAGE)).toLowerCase();
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/view/Gui.java:115]: I process words in lower case.

File: Ghiurutan V. Alexandru/Homework5/src/view/Gui.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		}
+		return description;
+	}
+
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/view/Gui.java:119]: Recursive method for adding words.Ensures that any word is either a word with description or not.

File: Ghiurutan V. Alexandru/Homework5/src/model/Dictionary.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+	@SuppressWarnings({ "unchecked" })
+	@Override
+	public void update(Observable o, Object arg) {
+		ArrayList<Word> deleteList = (ArrayList<Word>) arg;
+		for (Word word : deleteList) {
+			dictionary.remove(word);
+		}
+	}
+
```
Comment (AlexandruGH) [inline Ghiurutan V. Alexandru/Homework5/src/model/Dictionary.java:134]: Here the search logic is used.

File: Ghiurutan V. Alexandru/Homework5/src/model/Word.java (hunk 3 above)
Comment (ddelilah) [inline Ghiurutan V. Alexandru/Homework5/src/model/Word.java:8]: Observer can and should be used when any update on the dictionary is made. That way, the user is notified of the changes.

**Pull request merged**